From 622bf2463e930a9beecfdbd345ddff9b590a5589 Mon Sep 17 00:00:00 2001
From: andreas salvisberg <andreas.salvisberg@netmodule.com>
Date: Thu, 21 Nov 2013 14:45:39 +0100
Subject: [PATCH 18/20] zx3_ze7000: added configuration for skew values

Micrel Gbit phy on ZX3 module needs adaptions of the skew values
---
 board/netmodule/ze7000/board.c |   65 +++++++++++++++++++++++++++++++++++++++-
 include/configs/zx3_ze7000.h   |   10 +++----
 2 files changed, 69 insertions(+), 6 deletions(-)

diff --git a/board/netmodule/ze7000/board.c b/board/netmodule/ze7000/board.c
index 21a838d..c7c9a2f 100644
--- a/board/netmodule/ze7000/board.c
+++ b/board/netmodule/ze7000/board.c
@@ -32,6 +32,7 @@
 
 #include <nand.h>
 #include <i2c.h>
+#include <phy.h>
 
 #define BOOT_MODE_REG     (ZYNQ_SYS_CTRL_BASEADDR + 0x25C)
 #define BOOT_MODES_MASK    0x0000000F
@@ -39,7 +40,7 @@
 #define NOR_FLASH_MODE    (0x00000002)            /**< NOR  */
 #define NAND_FLASH_MODE   (0x00000004)            /**< NAND */
 #define SD_MODE           (0x00000005)            /**< Secure Digital card */
-#define JTAG_MODE	  (0x00000000)		  /**< JTAG */
+#define JTAG_MODE         (0x00000000)            /**< JTAG */
 
 extern int zynq_nand_init(struct nand_chip *nand_chip, int devnum);
 
@@ -220,6 +221,68 @@ void zx3_set_storage (int store) {
 }
 
 #ifdef CONFIG_CMD_NET
+int board_phy_config(struct phy_device *phydev)
+{
+	static int do_once;
+
+	if (!do_once) {
+		/* first interface, on module */
+		if (phydev->dev->iobase == ZYNQ_GEM_BASEADDR0) {
+			/* Giga skew value */
+			if (phydev->phy_id == 0x00221621) { /* KSZ9031 */
+				phy_write(phydev, phydev->addr, 0xD, 0x0002);
+				phy_write(phydev, phydev->addr, 0xE, 0x0008); // Reg 0x8
+				phy_write(phydev, phydev->addr, 0xD, 0x4002);
+				phy_write(phydev, phydev->addr, 0xE, 0x03FF); //3FF = max RXC and TXC delay
+			}
+			else {
+				printf ("unsupported PHY Model, ID:0x%08X\n", phydev->phy_id);
+				goto exit;
+			}
+		}
+		/* second interface, on board */
+		else if (phydev->dev->iobase == ZYNQ_GEM_BASEADDR1) {
+
+			char * devname = phydev->dev->name;
+
+			phy_write(phydev, phydev->addr, 0x16, 0x0002); /* Page 2 */
+			phy_write(phydev, phydev->addr, 0x15, 0x3010); /* GigaB */
+			phy_write(phydev, phydev->addr, 0x16, 0x0003); /* Page 3 */
+			phy_write(phydev, phydev->addr, 0x10, 0x1240); /* Led settings */
+			phy_write(phydev, phydev->addr, 0x11, 0x4425);
+
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR2, 0x16, 0x0002); /* Page 2 */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR2, 0x15, 0x3010); /* GigaB */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR2, 0x16, 0x0003); /* Page 3 */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR2, 0x10, 0x1240); /* Led settings */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR2, 0x11, 0x4425);
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR2, 0x00, 0x9140); /* Reset */
+
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR3, 0x16, 0x0002); /* Page 2 */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR3, 0x15, 0x3010); /* GigaB */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR3, 0x16, 0x0003); /* Page 3 */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR3, 0x10, 0x1240); /* Led settings */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR3, 0x11, 0x4425);
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR3, 0x00, 0x9140); /* Reset */
+
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR4, 0x16, 0x0002); /* Page 2 */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR4, 0x15, 0x3010); /* GigaB */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR4, 0x16, 0x0003); /* Page 3 */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR4, 0x10, 0x1240); /* Led settings */
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR4, 0x11, 0x4425);
+			miiphy_write (devname, CONFIG_ZYNQ_GEM_PHY_ADDR4, 0x00, 0x9140); /* Reset */
+		}
+	}
+
+	do_once = 1;
+exit:
+	/* call the standard PHY configuration as well */
+	if (phydev->drv->config)
+		phydev->drv->config(phydev);
+
+	return 0;
+}
+
 /**
  * Read the MAC address in the module EEPROM
  */
diff --git a/include/configs/zx3_ze7000.h b/include/configs/zx3_ze7000.h
index a641b80..a650c7b 100644
--- a/include/configs/zx3_ze7000.h
+++ b/include/configs/zx3_ze7000.h
@@ -28,11 +28,11 @@
 #undef CONFIG_ZYNQ_GEM1     //@@@as
 #define CONFIG_PHY_MARVEL
 #define CONFIG_PHY_MICREL
-#define CONFIG_ZYNQ_GEM_PHY_ADDR0        3
-#define CONFIG_ZYNQ_GEM_PHY_ADDR1        7
-#define CONFIG_ZYNQ_GEM_PHY_ADDR1_ADD1   6
-#define CONFIG_ZYNQ_GEM_PHY_ADDR1_ADD2   5
-#define CONFIG_ZYNQ_GEM_PHY_ADDR1_ADD3   4
+#define CONFIG_ZYNQ_GEM_PHY_ADDR0        3  /* Phy at GEM0 */
+#define CONFIG_ZYNQ_GEM_PHY_ADDR1        7  /* Phy at GEM1 */
+#define CONFIG_ZYNQ_GEM_PHY_ADDR2        6
+#define CONFIG_ZYNQ_GEM_PHY_ADDR3        5
+#define CONFIG_ZYNQ_GEM_PHY_ADDR4        4
 
 #define CONFIG_CMD_NET
 #define CONFIG_CMD_MII      /* mii utils */
-- 
1.7.10.4

